File: news_server/modules/source/infrastructure/getters/cankaoxiaoxi.ts
/**
 * 参考消息新闻源获取器
 * 从 newsnow/server/sources/cankaoxiaoxi.ts 迁移
 */
import type { SourceGetter } from '../../domain/getter/getter'
import { NewsItem } from '../../domain/newsitem/entity'

interface CankaoxiaoxiItem {
  data: {
    id: string
    title: string
    url: string
    publishTime: string
  }
}

interface CankaoxiaoxiResponse {
  list: CankaoxiaoxiItem[]
}

export const cankaoxiaoxiGetter: SourceGetter = async () => {
  const channels = ['zhongguo', 'guandian', 'gj']
  const responses = await Promise.all(
    channels.map((k) =>
      fetch(`https://china.cankaoxiaoxi.com/json/channel/${k}/list.json`)
    )
  )

  const results = await Promise.all(
    responses.map(async (response) => {
      if (!response.ok) {
        throw new Error(`Failed to fetch Cankaoxiaoxi: ${response.statusText}`)
      }
      return (await response.json()) as CankaoxiaoxiResponse
    })
  )

  const allItems = results
    .flatMap((res) => res.list)
    .map((k) =>
      NewsItem.fromEntity({
        id: k.data.id,
        title: k.data.title,
        url: k.data.url,
        extra: {
          date: k.data.publishTime,
        },
      })
    )
    .sort((m, n) => {
      const dateM = m.extra?.date ? new Date(m.extra.date as string).getTime() : 0
      const dateN = n.extra?.date ? new Date(n.extra.date as string).getTime() : 0
      return dateN - dateM
    })

  return allItems
}

File: modules/source/infrastructure/getters/cankaoxiaoxi.go
package getters

import "context"

func (r *Registry) cankaoxiaoxi(ctx context.Context) ([]Item, error) {
	channels := []string{"zhongguo", "guandian", "gj"}
	type row struct {
		Data struct {
			ID          string `json:"id"`
			Title       string `json:"title"`
			URL         string `json:"url"`
			PublishTime string `json:"publishTime"`
		} `json:"data"`
	}
	var out []Item
	for _, ch := range channels {
		var res struct {
			List []row `json:"list"`
		}
		if err := r.getJSON(ctx, "https://china.cankaoxiaoxi.com/json/channel/"+ch+"/list.json", nil, &res); err != nil {
			return nil, err
		}
		for _, k := range res.List {
			out = append(out, Item{ID: k.Data.ID, Title: k.Data.Title, URL: k.Data.URL, Extra: map[string]any{"date": k.Data.PublishTime}})
		}
	}
	return out, nil
}
